src/core/exception.d

```diff
@@ -563,25 +532,22 @@ extern (C) void onInvalidMemoryOperationError(void* pretend_sideffect = null) @t
 
 
 /**
- * A callback for switch errors in D.  A $(LREF SwitchError) will be thrown.
+ * A callback for switch errors in D.
  *
  * Params:
  *  file = The name of the file that signaled this error.
  *  line = The line number on which this error occurred.
- *
- * Throws:
- *  $(LREF SwitchError).
  */
 extern (C) void onSwitchError( string file = __FILE__, size_t line = __LINE__ ) @safe pure nothrow
 {
-    throw new SwitchError( file, line, null );
+    assert(0, "No appropriate switch clause found");
 }
 
 // Compiler lowers final switch default case to this (which is a runtime error)
 void __switch_errorT()(string file = __FILE__, size_t line = __LINE__) @trusted
 {
     // Consider making this a compile time check.
-    throw staticError!SwitchError(file, line, null);
+    assert(0, "No appropriate switch clause found");
```

Comment:
Another thing we need is the ability to test runtime assertions in `unittest`s.  Any ideas?

Reply:
It generally works just fine to catch an `AssertError` in a unit test. That being said, I don't really see anything to test here. The stuff that was in the `SwitchError` tests all had to do with testing that the members of `SwitchError` were initialized correctly. I guess that you could do the same with an `AssertError`, but unless `-unittest` makes `assert(0, ...)` still throw an `AssertError` when the assertion isn't directly in a `unittest` block when compiling with -release, then you couldn't test it in -release. And honestly, I don't think that testing that the exception is initialized with all of the right values is all that valuable. What you generally need to test is that it's thrown when it's supposed to and not thrown when it's not supposed to, and that would involve testing code that actually triggered `__switch_errorT()`. All around though, it does get pretty weird to try to test anything that involves `assert(0, ...)`, since pretty much by definition, that code is never supposed to be reached, and it's a serious bug in your program if it is.

Reply:
From the spec

“Unlike AssertExpressions used elsewhere, the assert is not assumed to hold, and upon assert failure the program is still in a defined state.”

“AssertExpression has different semantics if it is in a unittest or in contract.”

Reply:
I don't know if it's in the spec but it's been pointed out that `Error`s should not be caught. Since `assert` throws an `AssertError` it's not possible to create a unit test runner that continues after a failed test.

Reply:
Haha spec != implementation. It works quite well to catch an `AssertError` and lots of code in Phobos depends on this, e.g.

```
ag "collectException.*Error"
std/exception.d
168:    assert(collectExceptionMsg!AssertError(assertNotThrown!StringException(
176:    assert(collectExceptionMsg!AssertError(assertNotThrown!StringException(
180:    assert(collectExceptionMsg!AssertError(assertNotThrown!StringException(
184:    assert(collectExceptionMsg!AssertError(assertNotThrown!StringException(
311:    assert(collectExceptionMsg!AssertError(assertThrown!StringException(
706:        auto e = collectException!Error(enforceEx!OutOfMemoryError(false));
714:        auto e = collectException!Error(enforceEx!OutOfMemoryError(false, "file", 42));
770:    assert(collectException!RangeError(a[4], b));
1843:    assert(throwRangeError().ifThrown(0).collectException!RangeError() !is null);
1844:    assert(throwRangeError().ifThrown(e=>0).collectException!RangeError() !is null);
```

```
ag "catch.*Error"
std/typecons.d
4302:        catch (Error e) {}

std/bitmanip.d
708:    catch (AssertError ae)
712:    catch (AssertError ae)

std/datetime/package.d
293:        catch (AssertError e)
320:        catch (AssertError e)

std/range/package.d
720:    catch (AssertError unused)

std/exception.d
200:    catch (AssertError) assert(0);
206:    catch (AssertError) assert(0);
212:    catch (AssertError) assert(0);
218:    catch (AssertError) assert(0);
227:        catch (AssertError) thrown = true;
238:        catch (AssertError) thrown = true;
249:        catch (AssertError) thrown = true;
261:        catch (AssertError) thrown = true;
327:    catch (AssertError) assert(0);
334:    catch (AssertError) assert(0);
341:    catch (AssertError) assert(0);
349:    catch (AssertError) assert(0);
356:        catch (AssertError)
366:        catch (AssertError)
376:        catch (AssertError)
386:        catch (AssertError)

std/experimental/allocator/gc_allocator.d
69:        catch (OutOfMemoryError)

std/parallelism.d
4830:    catch (ParallelForeachError e) {}
```

Reply:
> Haha spec != implementation. It works quite well to catch an AssertError and lots of code in Phobos depends on this, e.g

Yeah, I know. I do that as well. But I would feel better if it was an Exception instead of an Error.